File: recursion/03_Tower_Of_Hanoi/03_Tower_Of_Hanoi.js
const testCases = [{ n: 1 }, { n: 2 }, { n: 3 }, { n: 4 }];

(function () {
  testCases.forEach((testcase, i) => {
    console.log(`n= ${testcase.n}`);
    console.log(`testcase ${i}:  ${toh(testcase.n, 1, 3, 2)}`);
  });
})();

// Approach : T O(n^2) | S O(n)
function toh(N, from, to, aux) {
  let count = 0;

  function helper(N, from, to, aux) {
    if (N === 1) {
      count += 1;
      console.log("move disk " + N + " from rod " + from + " to rod " + to);
      return;
    }
    helper(N - 1, from, aux, to);
    count += 1;
    console.log("move disk " + N + " from rod " + from + " to rod " + to);
    helper(N - 1, aux, to, from);
  }

  helper(N, from, to, aux);
  return count;
}

File: recursion/03_Tower_Of_Hanoi/go/03_Tower_Of_Hanoi.go
package main

import "fmt"

func main() {
	n := 3
	totalMoves := toh(n, 1, 3, 2)
	fmt.Printf("Total moves for %d disks: %d\n", n, totalMoves)
	
	n = 2
	totalMoves = toh(n, 1, 3, 2)
	fmt.Printf("Total moves for %d disks: %d\n", n, totalMoves)
}

// Approach : T O(n^2) | S O(n)
func toh(n int, from int, to int, aux int) int {
	if n == 1 {
		fmt.Printf("Move disk 1 from rod %d to rod %d\n", from, to)
		return 1
	}
	moves := 0
	moves += toh(n-1, from, aux, to)
	fmt.Printf("Move disk %d from rod %d to rod %d\n", n, from, to)
	moves++
	moves += toh(n-1, aux, to, from)
	return moves
}
